pkg/prove/proof.go

```diff
@@ -0,0 +1,187 @@
+package prove
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/celestiaorg/rsmt2d"
+	tmbytes "github.com/tendermint/tendermint/libs/bytes"
+	"github.com/tendermint/tendermint/pkg/consts"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"github.com/tendermint/tendermint/types"
+)
+
+const (
+	usableShareSize = consts.ShareSize - consts.NamespaceSize - consts.ShareReservedBytes
+)
+
+// TxInclusion uses the provided block data to progressively generate rows
+// of a data square, and then using those shares to creates nmt inclusion proofs
+// It is possible that a transaction spans more than one row. In that case, we
+// have to return two proofs.
+func TxInclusion(codec rsmt2d.Codec, data types.Data, origSquareSize, txIndex int) (types.TxProof, error) {
+	startPos, endPos := txSharePosition(data.Txs, txIndex)
+	startRow := startPos / origSquareSize
+	endRow := endPos / origSquareSize
+
+	if (endPos - startPos) > 1 {
+		return types.TxProof{}, errors.New("transaction spanned more than two shares, this is not yet supported")
+	}
+
+	rowShares := genRowShares(consts.DefaultCodec(), data, origSquareSize, startRow, endRow)
+
+	var proofs []*tmproto.NMTProof  //nolint:prealloc // rarely will this contain more than a single proof
+	var shares [][]byte             //nolint:prealloc // rarely will this contain more than a single share
+	var rowRoots []tmbytes.HexBytes //nolint:prealloc // rarely will this contain more than a single root
+	for i, row := range rowShares {
+		// create an nmt to use to generate a proof
+		tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(origSquareSize))
+		for j, share := range row {
+			tree.Push(
+				share,
+				rsmt2d.SquareIndex{
+					Axis: uint(i),
+					Cell: uint(j),
+				},
+			)
+		}
+
+		var pos int
+		if i == 0 {
+			pos = startPos - (startRow * origSquareSize)
+		} else {
+			pos = endPos - (endRow * origSquareSize)
+		}
+
+		shares = append(shares, row[pos])
+
+		proof, err := tree.Prove(pos)
+		if err != nil {
+			return types.TxProof{}, err
```

Comment:
Is there a reason to return an empty proof instead of `nil` here?

Reply:
nil would be nicer, but then we have to return a pointer to a `TxProof`

Reply:
Ah I see